### PR TITLE
Search by GitHub attributes on non-roster AssignmentRepos

### LINKED ISSUE #1978: Allow teachers to search and sort by GitHub handles
### Feature request :sparkles:
We now have caching for GitHub handles. So we should now allow teachers to search and sort by GitHub handles. So we can now rollout this feature to teachers that do not have rosters.

### PR DESCRIPTION
**Pull request by thebradbain**
## What
Closes #1978 

Because we now cache GitHub handles, we can now enable teachers to search and sort by GitHub login for classroom organizations that do not have a roster linked to them.

This PR is also an opportunity to DRY the search functionality into a model concern (`Searchable`) much like the sort functionality (`Sortable`) already was.

### CODE REVIEW

File: app/controllers/assignments_controller.rb [hunk 1]
```diff
@@ -129,8 +129,12 @@ def set_assignment
     @assignment = @organization.assignments.includes(:assignment_invitation).find_by!(slug: params[:id])
   end
 
+  def set_list_type
+    @list_type = @organization.roster ? :roster_entries : :assignment_repos
```
Comment (thebradbain) [inline app/controllers/assignments_controller.rb]: I've created a `@list_type` instance variable set in a `before_action` to help make the view logic a little more readable, whereas previously were only relying `if @organization.roster` scattered everywhere for the same functionality

**Comment by thebradbain** [comment]
As an aside, you'll notice the actual backend changes to enable search/sort on unlinked assignment repos is pretty trivial and straightforward, _but_ the view logic as it currently exists, while ultimately trivial changes, is not very DRY. This is perhaps an indication that we should consider splitting up `AssignmentsController#show` into two actions depending on whether or not  `@organization.roster` is present for the assignment, because the views are differing enough now that scattered `if/else` statements are beginning to be insufficient.

**Comment by d12** [comment]
> This is perhaps an indication that we should consider splitting up AssignmentsController#show into two actions depending on whether or not @organization.roster is present for the assignment, because the views are differing enough now that scattered if/else statements are beginning to be insufficient.

+ 100, I'd like to redesign these views to make more logical sense, you can tell that we're overcomplicating the design by the horrible spaghetti code we have. I mention it a little bit here: https://github.com/education/classroom/issues/1969#issuecomment-509711272

Also a note, lets not merge this until I populate the caches. Right now, the cache get populated when `github_user.login` gets fetched. That means some users have logins, but a lot still don't. I'll write a task to go ahead and backfill them all, and make sure we're filling in the login on creation.

**Comment by d12** [comment]
Issue to populate the caches: https://github.com/education/classroom/issues/1984

**Comment by thebradbain** [comment]
@d12 The tests have been added so whenever the cache is warmed we should be ready to rollout

**Comment by d12** [comment]
Thanks @thebradbain ! I have something working locally, just gonna do a bit of local testing and then I'll push that up. The task may take a very long time to run (days), we'll see 😬

**Pull request merged**